irods/test/ticket_test.py

```diff
@@ -79,6 +79,30 @@ def tearDown(self):
             for u in self.users:
                 ses.users.remove(u)
 
+    def test_session_clone_initialization_of_tickets_applied__issue_619(self):
+        ticket_string=''
+        try:
+            # Make a ticket with target collection being a rodsuser's home:
+            with self.login(self.alice) as alice:
+                alice_home_path = helpers.home_collection(alice)
+                ticket_string = Ticket(alice).issue('read', alice_home_path).string
+
+            # Create an original session. Cause a connection to be generated and used.
+            old_sess = helpers.make_session()
+            Ticket(old_sess, ticket_string).supply()
+            old_sess.collections.get(alice_home_path)
+
+            # Clone the session.
+            new_sess = old_sess.clone()
+
+            # Original session object should have had the ticket applied to a connection.
+            self.assertLess(0, len(old_sess.ticket_applied))
+            # Cloned session object should be starting out with a zero-length ticket/connection map.
+            self.assertEqual(0, len(new_sess.ticket_applied))
```

Comment:
Based on this test, the new session does not inherit the ticket.

If that's true, consider adding an assertion that proves the new session cannot access the home collection of alice.

Reply:
It does inherit the ticket, in that the[ ticket string is copied into](https://github.com/irods/python-irodsclient/blob/527417da41a997198e514e0e211b5422585a97a0/irods/session.py#L177) in the ticket__ attribute of the cloned session.  So the new session will in fact be able to access alice's home directory.

Unless we were to clone with : `new_sess  = old_sess.clone( ticket = '')`

Reply:
let's add an assertion to make it known to the next reader.

Reply:
Can do.  @trel - assume you mean that we assert alice's home collection *is* accessible by the cloned session?

Reply:
> Can do. @trel - assume you meant that we assert alice's home collection _is_ accessible by the cloned session?

Yes, make an assertion, label it well (read: why this matters), and we're good.

Reply:
For my own understanding...

Right now, we are asserting that the ticket is marked as "not applied" in the newly cloned session even though it is marked as "applied" in the source session which was cloned.

We would like to assert that the ticket is marked as "not applied" in the newly cloned session **and** that it can in fact be used **and** that none of this has any effect on the source session which was cloned.

Reply:
@alanking - The ticket was applied already, within the old session, to an individual server connection.  Although the new session inherits the old one's associated ticket , it only **_applies_** that to a given connection  once it's made.  The already-applied tickets are tracked by connection, but the new session has no connections yet.
This is all by design since session cloning was originally designed to support multi-1247 i/o, which must in turn be able to support tickets.
So yes, I'm already asserting the ticket is not yet applied in the new session, but has been applied in the old one.

Reply:
Okay, thanks for the explanation.

Reply:
> I think so. Adding an assertion and comment for that would clear things up.
> 
> My confusion stems from the `assertLess`, `assertEqual`, and `ticket_applied`. Please explain how those should be interpreted.

Sorry didn't see this before.  Each session is a pool capable of managing several connections.  Since tickets must be applied within each server socket connection (of class `irods.connection.Connection`), this ticket_applied data structure is used to track the application of the ticket to those individual server connections.

When a session object is cloned, the new copy does not contain any such connections, because the creation of connections won't occur until an operation is performed using the new session.  Thus, just after creation time, the cloned session will have applied no tickets to any connections (hence the assertion:  `self.assertEqual(0, len(new_sess.ticket_applied))` ).  However, the original session from which the clone was made will already have applied the ticket to at least one connection within its pool, since we invoked sess.collections.get(path) on it; hence the assertion:`self.assertLess(0, len(old_sess.ticket_applied))`

Reply:
> > Can do. @trel - assume you meant that we assert alice's home collection _is_ accessible by the cloned session?
> 
> Yes, make an assertion, label it well (read: why this matters), and we're good.

Ok I believe I've done this.  Assertion is in this case implicit, since its failing would raise an Exception and cause the overall test to fail.

Reply:
Looks good to me.

Resolving.